File: apps/mobile/test/booking_status_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:snip_mobile/models/booking.dart';

bool isValidBookingTransition(BookingStatus from, BookingStatus to) {
  if (from == to) return true;
  const allowed = <BookingStatus, Set<BookingStatus>>{
    BookingStatus.pending: {
      BookingStatus.confirmed,
      BookingStatus.cancelled,
    },
    BookingStatus.confirmed: {
      BookingStatus.checkedIn,
      BookingStatus.cancelled,
      BookingStatus.noShow,
    },
    BookingStatus.checkedIn: {
      BookingStatus.inProgress,
      BookingStatus.cancelled,
      BookingStatus.noShow,
    },
    BookingStatus.inProgress: {
      BookingStatus.completed,
      BookingStatus.cancelled,
    },
  };
  return allowed[from]?.contains(to) ?? false;
}

void main() {
  test('allows confirmed to checked_in', () {
    expect(
      isValidBookingTransition(BookingStatus.confirmed, BookingStatus.checkedIn),
      isTrue,
    );
  });

  test('blocks completed to confirmed', () {
    expect(
      isValidBookingTransition(BookingStatus.completed, BookingStatus.confirmed),
      isFalse,
    );
  });
}

File: apps/web/tests/booking-status.test.ts
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import {
  assertBookingTransition,
  isValidBookingTransition,
} from "../lib/booking/status";

describe("booking status transitions", () => {
  it("allows confirmed → checked_in", () => {
    assert.equal(isValidBookingTransition("confirmed", "checked_in"), true);
  });

  it("blocks completed → confirmed", () => {
    assert.equal(isValidBookingTransition("completed", "confirmed"), false);
  });

  it("allows same-status no-op", () => {
    assert.equal(isValidBookingTransition("in_progress", "in_progress"), true);
  });

  it("throws on invalid assert", () => {
    assert.throws(() => assertBookingTransition("cancelled", "confirmed"));
  });

  it("allows happy-path check-in workflow", () => {
    const path = [
      ["confirmed", "checked_in"],
      ["checked_in", "in_progress"],
      ["in_progress", "completed"],
    ] as const;
    for (const [from, to] of path) {
      assert.equal(isValidBookingTransition(from, to), true);
    }
  });
});
